File: src/frontend/services/api_client.py
"""API client for backend communication."""

import logging
from typing import Any

import httpx
from httpx import AsyncClient, HTTPStatusError

from config import settings

logger = logging.getLogger(__name__)


class APIError(Exception):
    """API communication errors."""

    pass


class BackendAPIClient:
    """Client for communicating with the FastAPI backend."""

    def __init__(self, access_token: str | None = None):
        """Initialize the API client.

        Args:
            access_token: Bearer token for authentication
        """
        self.base_url = settings.backend_url.rstrip("/")
        self.access_token = access_token
        self.timeout = httpx.Timeout(30.0)

    def _get_headers(self) -> dict[str, str]:
        """Get HTTP headers for API requests.

        Returns:
            Headers dictionary with authentication and content type
        """
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            "User-Agent": "Azure-DevOps-AI-Agent-Frontend/1.0",
        }

        if self.access_token:
            headers["Authorization"] = f"Bearer {self.access_token}"

        return headers

    async def send_chat_message(
        self,
        message: str,
        conversation_id: str | None = None,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Send a chat message to the AI agent.

        Args:
            message: User message text
            conversation_id: Optional conversation identifier
            context: Optional context information

        Returns:
            AI response with message and metadata
        """
        try:
            payload = {
                "message": message,
                "conversation_id": conversation_id,
                "context": context or {},
            }

            async with AsyncClient(timeout=self.timeout) as client:
                response = await client.post(
                    f"{self.base_url}/chat/message",
                    json=payload,
                    headers=self._get_headers(),
                )
                response.raise_for_status()

                result = response.json()
                logger.info(
                    f"Chat message sent successfully, got response: {len(result.get('data', {}).get('response', ''))} chars"
                )
                return result

        except HTTPStatusError as e:
            logger.error(
                f"HTTP error sending chat message: {e.response.status_code} - {e.response.text}"
            )
            raise APIError(f"Chat API error: {e.response.status_code}") from e
        except Exception as e:
            logger.error(f"Error sending chat message: {e}")
            raise APIError(f"Failed to send chat message: {e}") from e

    async def get_conversations(self, limit: int = 20) -> list[dict[str, Any]]:
        """Get user's conversation history.

        Args:
            limit: Maximum number of conversations to retrieve

        Returns:
            List of conversation objects
        """
        try:
            params = {"limit": limit}

            async with AsyncClient(timeout=self.timeout) as client:
                response = await client.get(
                    f"{self.base_url}/chat/conversations",
                    params=params,
                    headers=self._get_headers(),
                )
                response.raise_for_status()

                result = response.json()
                conversations = result.get("data", [])
                logger.info(f"Retrieved {len(conversations)} conversations")
                return conversations

        except HTTPStatusError as e:
            logger.error(f"HTTP error getting conversations: {e.response.status_code}")
            raise APIError(f"Conversations API error: {e.response.status_code}") from e
        except Exception as e:
            logger.error(f"Error getting conversations: {e}")
            raise APIError(f"Failed to get conversations: {e}") from e

    async def get_conversation(self, conversation_id: str) -> dict[str, Any]:
        """Get a specific conversation with its message history.

        Args:
            conversation_id: Conversation identifier

        Returns:
            Conversation object with messages
        """
        try:
            async with AsyncClient(timeout=self.timeout) as client:
                response = await client.get(
                    f"{self.base_url}/chat/conversations/{conversation_id}",
                    headers=self._get_headers(),
                )
                response.raise_for_status()

                result = response.json()
                logger.info(f"Retrieved conversation {conversation_id}")
                return result.get("data", {})

        except HTTPStatusError as e:
            logger.error(f"HTTP error getting conversation: {e.response.status_code}")
            raise APIError(f"Conversation API error: {e.response.status_code}") from e
        except Exception as e:
            logger.error(f"Error getting conversation: {e}")
            raise APIError(f"Failed to get conversation: {e}") from e

    async def health_check(self) -> dict[str, Any]:
        """Check backend API health status.

        Returns:
            Health status information
        """
        try:
            async with AsyncClient(timeout=httpx.Timeout(5.0)) as client:
                response = await client.get(
                    f"{self.base_url}/health", headers={"Accept": "application/json"}
                )
                response.raise_for_status()

                result = response.json()
                logger.info("Backend health check successful")
                return result

        except HTTPStatusError as e:
            logger.error(f"Backend health check failed: {e.response.status_code}")
            raise APIError(f"Backend health check failed: {e.response.status_code}") from e
        except Exception as e:
            logger.error(f"Backend health check error: {e}")
            raise APIError(f"Backend health check failed: {e}") from e

    async def get_user_profile(self) -> dict[str, Any]:
        """Get authenticated user's profile information.

        Returns:
            User profile data
        """
        try:
            async with AsyncClient(timeout=self.timeout) as client:
                response = await client.get(
                    f"{self.base_url}/auth/profile", headers=self._get_headers()
                )
                response.raise_for_status()

                result = response.json()
                logger.info("Retrieved user profile")
                return result.get("data", {})

        except HTTPStatusError as e:
            logger.error(f"HTTP error getting user profile: {e.response.status_code}")
            raise APIError(f"Profile API error: {e.response.status_code}") from e
        except Exception as e:
            logger.error(f"Error getting user profile: {e}")
            raise APIError(f"Failed to get user profile: {e}") from e

    def set_access_token(self, access_token: str) -> None:
        """Update the access token for API requests.

        Args:
            access_token: New bearer token
        """
        self.access_token = access_token
        logger.info("Access token updated")

    def clear_access_token(self) -> None:
        """Clear the access token."""
        self.access_token = None
        logger.info("Access token cleared")

File: src/frontend/src/services/api-client.ts
/**
 * API client for communicating with the FastAPI backend.
 * Uses the new client config API instead of NEXT_PUBLIC_* environment variables.
 */

import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
import { getCachedClientConfig } from '@/hooks/use-client-config';
import { trackApiCall, trackException } from '@/lib/telemetry';
import { MfaHandler, MfaHandlerOptions } from '@/services/mfa-handler';
import type {
  ApiResponse,
  ChatRequest,
  ChatResponse,
  BackendStatus,
} from '@/types';

// Extend AxiosRequestConfig to include metadata
declare module 'axios' {
  interface InternalAxiosRequestConfig {
    metadata?: {
      startTime: number;
    };
  }
}

export class ApiClient {
  private client: AxiosInstance;
  private accessToken: string | null = null;
  private mfaHandler: MfaHandler | null = null;

  constructor() {
    // Initialize axios client with a placeholder URL
    // The actual baseURL will be set lazily when first request is made
    this.client = axios.create({
      timeout: 30000,
      headers: {
        'Content-Type': 'application/json',
      },
    });

    // Request interceptor to add auth token and ensure baseURL is set
    this.client.interceptors.request.use(
      (config) => {
        // Lazily set the baseURL if not already set
        if (!config.baseURL) {
          const clientConfig = getCachedClientConfig();
          if (clientConfig) {
            config.baseURL = clientConfig.backend.url;
          } else {
            // Fallback for when config isn't loaded yet
            config.baseURL = 'http://localhost:8000';
          }
        }

        if (this.accessToken) {
          config.headers.Authorization = `Bearer ${this.accessToken}`;
        }
        return config;
      },
      (error) => {
        return Promise.reject(error);
      }
    );

    // Response interceptor for logging and error handling
    this.client.interceptors.response.use(
      (response: AxiosResponse) => {
        // Track successful API calls
        const startTime = response.config.metadata?.startTime || Date.now();
        const duration = Date.now() - startTime;

        trackApiCall(
          response.config.url || '',
          response.config.method?.toUpperCase() || 'GET',
          response.status,
          duration,
          true
        );

        return response;
      },
      async (error) => {
        // Track failed API calls
        const response = error.response;
        const startTime = error.config?.metadata?.startTime || Date.now();
        const duration = Date.now() - startTime;

        trackApiCall(
          error.config?.url || '',
          error.config?.method?.toUpperCase() || 'GET',
          response?.status || 0,
          duration,
          false
        );

        // Handle MFA challenge if we have a handler configured and it's an MFA error
        if (this.mfaHandler && this.mfaHandler.isMfaChallengeError(error)) {
          try {
            console.log('MFA challenge detected, attempting to handle...');
            const newToken = await this.mfaHandler.handleMfaChallengeFromError(error);
            
            // Update our token and retry the request
            this.setAccessToken(newToken);
            
            // Clone and retry the original request with the new token
            const originalRequest = { ...error.config };
            originalRequest.headers = originalRequest.headers || {};
            originalRequest.headers.Authorization = `Bearer ${newToken}`;
            
            return this.client.request(originalRequest);
          } catch (mfaError) {
            console.error('Failed to handle MFA challenge:', mfaError);
            // Return original error if MFA handling fails
          }
        }

        // Track exception for serious errors
        if (!response || response.status >= 500) {
          trackException(error, {
            url: error.config?.url,
            method: error.config?.method,
            status: response?.status,
          });
        }

        return Promise.reject(error);
      }
    );

    // Add request timestamp for duration tracking
    this.client.interceptors.request.use((config) => {
      config.metadata = { startTime: Date.now() };
      return config;
    });
  }

  /**
   * Set the MFA handler for automatic MFA challenge handling
   */
  setMfaHandler(mfaHandler: MfaHandler) {
    this.mfaHandler = mfaHandler;
  }

  /**
   * Set the access token for authenticated requests
   */
  setAccessToken(token: string | null) {
    this.accessToken = token;
  }

  /**
   * Set the backend API token for authenticated requests (preferred method)
   * This method is specifically for backend API tokens as part of the split token approach
   */
  setBackendApiToken(token: string | null) {
    this.accessToken = token;
  }

  /**
   * Get backend health status
   */
  async getHealthStatus(): Promise<BackendStatus> {
    try {
      const response = await this.client.get<BackendStatus>('/health');
      return response.data;
    } catch (error) {
      console.error('Health check failed:', error);
      return {
        status: 'unhealthy',
        services: {},
      };
    }
  }

  /**
   * Send a chat message
   */
  async sendMessage(request: ChatRequest): Promise<ApiResponse<ChatResponse>> {
    try {
      const response = await this.client.post<ChatResponse>(
        '/chat/message',
        request
      );
      return {
        data: response.data,
        success: true,
      };
    } catch (error: any) {
      const errorMessage =
        error.response?.data?.detail ||
        error.message ||
        'Failed to send message';
      return {
        error: errorMessage,
        success: false,
      };
    }
  }

  /**
   * Get chat history
   */
  async getChatHistory(
    conversationId?: string
  ): Promise<ApiResponse<ChatResponse[]>> {
    try {
      const params = conversationId ? { conversation_id: conversationId } : {};
      const response = await this.client.get<ChatResponse[]>('/chat/history', {
        params,
      });
      return {
        data: response.data,
        success: true,
      };
    } catch (error: any) {
      const errorMessage =
        error.response?.data?.detail ||
        error.message ||
        'Failed to get chat history';
      return {
        error: errorMessage,
        success: false,
      };
    }
  }

  /**
   * Get user projects from Azure DevOps
   */
  async getProjects(): Promise<ApiResponse<any[]>> {
    try {
      const response = await this.client.get<any[]>('/projects');
      return {
        data: response.data,
        success: true,
      };
    } catch (error: any) {
      const errorMessage =
        error.response?.data?.detail ||
        error.message ||
        'Failed to get projects';
      return {
        error: errorMessage,
        success: false,
      };
    }
  }

  /**
   * Generic GET request
   */
  async get<T>(
    url: string,
    config?: AxiosRequestConfig
  ): Promise<ApiResponse<T>> {
    try {
      const response = await this.client.get<T>(url, config);
      return {
        data: response.data,
        success: true,
      };
    } catch (error: any) {
      const errorMessage =
        error.response?.data?.detail || error.message || 'Request failed';
      return {
        error: errorMessage,
        success: false,
      };
    }
  }

  /**
   * Generic POST request
   */
  async post<T>(
    url: string,
    data?: any,
    config?: AxiosRequestConfig
  ): Promise<ApiResponse<T>> {
    try {
      const response = await this.client.post<T>(url, data, config);
      return {
        data: response.data,
        success: true,
      };
    } catch (error: any) {
      const errorMessage =
        error.response?.data?.detail || error.message || 'Request failed';
      return {
        error: errorMessage,
        success: false,
      };
    }
  }
}

// Global API client instance
export const apiClient = new ApiClient();

// Export default for easier imports
export default apiClient;
